### DISCUSSION TITLE
MongoDB exceeding 500 connection limit

### DISCUSSION THREAD

**Discussion by rawcomposition** [vercel/community#1211] [Help] [answered]
I'm using Nextjs, Mongoose with Atlas free tier, and deploying to Vercel. I've started having a burst in MongoDB connections resulting in emails from Atlas saying I've exceeded my limit of 500 connections. This coincides with all my Nextjs API endpoints and pages that run MongoDB queries to reach the 10s timeout limit. Normally, those functions take < 200ms to finish. This pattern happens about once a day, usually early in the morning, and lasts about 1 to 2 minutes until the MongoDB connections go down.

I followed the `with-mongodb-mongoose` official Nextjs example for caching connections. This seems to be working, I can see the functions console.logging successfully connecting, and future requests run without attempting to connect.

You can see my connection code here: https://github.com/rawcomposition/birdinghotspots/blob/main/lib/mongo.ts#L12-L56

I've attempted setting the MongoDB `maxPoolSize` option in Mongoose to 10, and this had no affect whatsoever. My average connections at any given point are bout 40-80, when it's not surging.

I'm guessing my Vercel functions are timing out because I reach the 500 connection limit and it blocks queries for a few minutes, but I'm not certain that's the problem. Regardless, if I can figure out why my connection counts are surging, I bet this problem would resolve. You can see it successfully connects

I've asked on the MongoDB forums and got no response at all. Any advice or tips on how to debug this or settings I could change to make Mongoose/MongoDB behave better in a serverless environment?

<img width="852" alt="Screen Shot 2023-01-04 at 11 45 22 AM" src="https://user-images.githubusercontent.com/740090/210676728-7d16d04f-430b-4bf0-ada6-840d737a9673.png">

And just to prove that the connections seem to be caching normally, see below, there's no connection messages logged:
<img width="1267" alt="Screen Shot 2023-01-04 at 4 47 25 PM" src="https://user-images.githubusercontent.com/740090/210677105-3cd62672-ba90-4181-82bd-bdc63f262a53.png">

**Comment by DanKamNdi** [comment, marked as answer]
Find a way of closing the connection to MongoDB. A new connection is made every time the connect function is called. To fix this you need to close the connection after you are done performing the operations you are doing on the database.

**Reply by josealvarez97** [reply]
@DanKamNdi Do you have a code example?

**Comment by josealvarez97** [comment]
@rawcomposition If you found a solution, could you help by sharing a code example?

**Comment by josealvarez97** [comment]
@amyegan Is there a code example about how to approach this with NextJS and Mongoose? Thank you.

This seems to be a common issue https://www.mongodb.com/community/forums/t/atlas-next-js-now-connection-issues/16217/13.

**Reply by amyegan** [reply]
This kind of problem is common with traditional databases and serverless environments in general. Automatic scaling of functions during traffic spikes makes it more likely that you'll exhaust available database connections. 

This guide explains best practices for connecting to relational databases with Serverless: https://vercel.com/guides/connection-pooling-with-serverless-functions

> Traditional relational databases were built for long-running compute instances, not the ephemeral nature of serverless functions. Serverless Functions are stateless and asynchronous. They are not designed for persistent connections to a database.
— [Connection Pooling with Serverless Functions: Connecting to Databases](https://vercel.com/guides/connection-pooling-with-serverless-functions#connecting-to-databases)

Some solutions, like Atlas, are designed for serverless apps. Services optimized for serverless connections generally expose an API layer on top of the underlying database. I know there is a [MongoDB Atlas integration](https://vercel.com/integrations/mongodbatlas) and related [MongoDB Starter](https://vercel.com/templates/next.js/mongodb-starter) that can make things easier if you were using Atlas. 

>Depending on your data workload, you might explore using other storage solutions that don’t require persistent connections. For example [DynamoDB](https://vercel.com/templates/next.js/aws-dynamodb-with-nextjs-api-routes).
— [Connection Pooling with Serverless Functions: Other Data Storage Solutions](https://vercel.com/guides/connection-pooling-with-serverless-functions#other-data-storage-solutions)

**Comment by rawcomposition** [comment]
@josealvarez97 there's more discussion on the MongoDB Forum here that I started: https://www.mongodb.com/community/forums/t/large-number-of-connections-with-mongoose-and-vercel/204917/5

I also had some extensive private dialog with Alex. In the end, he suggested setting the `maxIdleTimeMS=60000` value. This evidently isn't recommended for all scenarios, but between this and them evidently making some adjustments to Atlas on their end, my issues largely cleared up. Every few months I still get the error about nearing 500 connections, but far fewer, and I'm not seeing the mass timeout errors I was noticing before.

When I asked Alex if the `maxIdleTimeMS` is recommneded for all serverless setups, he said this:

> This guidance has proven to be effective for Google Cloud Functions, and in some instances AWS Lambda however we are not promoting this yet as general guidance for all serverless environments. Under the covers maxIdleTimeMS is a signal to the connection pool that a connection should be marked as idle once the threshold is met - causing it to be closed/recreated on the next connection checkout event. maxIdleTimeMS has a default of 0, which means “allow these connections to idle forever”, so the “cost” to the application is that if connections idle for a while they’ll be closed and recreated.

Early on I did make the adjustment to my connection logic per DanKamNdi's suggestion. I'm not sure if it helped much, but it certainly didn't hurt. I followed this example here: https://github.com/vercel/next.js/blob/canary/examples/with-mongodb-mongoose/lib/dbConnect.js
